StringImpl should construct a valid span to pass to StringImplShape

squelart opened the pull request:
#### bc6d37008fb3c4858a96c78fa656af0a2ec9f6da
<pre>
StringImpl should construct a valid span to pass to StringImplShape
<a href="https://bugs.webkit.org/show_bug.cgi?id=289518">https://bugs.webkit.org/show_bug.cgi?id=289518</a>
<a href="https://rdar.apple.com/146724107">rdar://146724107</a>

Reviewed by Geoffrey Garen.

StringImpl&apos;s constructor passes a span to StringImplShape, with a
nullptr data and non-null size.
It&apos;s only used to initialize StringImplShape members, the data pointer
is soon overriden with a valid pointer, so there&apos;s no UB per se.
However, it&apos;s a dangerous state that may trigger assertions in some
libraries, e.g.:
<a href="https://github.com/llvm/llvm-project/blob/1e83d975d72037567afd9d3b22bb063b442ec045/libcxx/include/span#L273">https://github.com/llvm/llvm-project/blob/1e83d975d72037567afd9d3b22bb063b442ec045/libcxx/include/span#L273</a>

This patch moves some code from the constructor function body, into a
static member function template that directly constructs the final
valid span given to StringImplShape.

* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::toStringImplMallocSpan):
(WTF::StringImpl::StringImpl):

Canonical link: <a href="https://commits.webkit.org/292384@main">https://commits.webkit.org/292384@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a26fc869310b21a8eddbbb349f30eec7e480275d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95780 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5309 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100834 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46285 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15677 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23821 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73040 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30285 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98783 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11738 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86509 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53374 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11465 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4244 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45624 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88448 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81666 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4351 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102865 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94400 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22839 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16672 "Found 2 new test failures: html5lib/generated/run-entities01-write.html imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/no-coop-coep.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82074 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23091 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82540 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81439 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20438 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26019 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3479 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16175 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22806 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27967 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/117879 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22465 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25941 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -930,18 +931,22 @@ inline StringImpl::StringImpl(unsigned length)
-inline StringImpl::StringImpl(MallocSpan<LChar, Malloc> characters)
-    : StringImplShape(s_refCountIncrement, { static_cast<const LChar*>(nullptr), characters.span().size() }, s_hashFlag8BitBuffer | StringNormal | BufferOwned)
+template<typename CharacterType, typename Malloc>
+std::span<const CharacterType> StringImpl::ensureStringImplMallocSpan(MallocSpan<CharacterType, Malloc>&& characters)
```
We usually use `ensure` to indicate that we're doing one-time initialization of a persistent value. I would call this `toStringImplMallocSpan`.

geoffreygaren commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -930,18 +931,22 @@ inline StringImpl::StringImpl(unsigned length)
 {
     if constexpr (std::is_same_v<Malloc, StringImplMalloc>)
-        m_data8 = characters.leakSpan().data();
+        return characters.leakSpan();
```
We can move-construct a MallocSpan return value here. That's a little clearer for a conversion function. (Then our StringImpl caller can leak the result, since StringImpl promises to delete in its destructor.)

squelart replied:
Makes sense. thank you for the thoughtful review.

geoffreygaren commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -930,18 +931,22 @@ inline StringImpl::StringImpl(unsigned length)
+        auto buffer = MallocSpan<CharacterType, StringImplMalloc>::malloc(characters.sizeInBytes());
         copyCharacters(buffer.mutableSpan(), characters.span());
-        m_data8 = buffer.leakSpan().data();
+        return buffer.leakSpan();
```
Ditto.

geoffreygaren reviewed (approved):
r=me with some suggestions for how to clarify lifetime.

squelart committed:
StringImpl should construct a valid span to pass to StringImplShape
https://bugs.webkit.org/show_bug.cgi?id=289518
rdar://146724107

Reviewed by Geoffrey Garen.

StringImpl's constructor passes a span to StringImplShape, with a
nullptr data and non-null size.
It's only used to initialize StringImplShape members, the data pointer
is soon overriden with a valid pointer, so there's no UB per se.
However, it's a dangerous state that may trigger assertions in some
libraries, e.g.:
https://github.com/llvm/llvm-project/blob/1e83d975d72037567afd9d3b22bb063b442ec045/libcxx/include/span#L273

This patch moves some code from the constructor function body, into a
static member function template that directly constructs the final
valid span given to StringImplShape.

* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::toStringImplMallocSpan):
(WTF::StringImpl::StringImpl):

Canonical link: https://commits.webkit.org/292384@main
